src/vm-repair/azext_vm_repair/custom.py

```
@@ -102,6 +131,50 @@ def create(cmd, vm_name, resource_group_name, repair_password=None, repair_usern
             _call_az_command(attach_disk_command)
+
+            # Install extension in case of single pass encrypted VM
+            try:
+                if encryption_type == "single_with_kek":
+                    encryption_type, key_vault, key_encryption_url = _uses_encrypted_disk(source_vm)
+
+                    install_ade_extension_command = 'az vm encryption enable --disk-encryption-keyvault {vault} --name {repair} --resource-group {g} --key-encryption-key {kek_url} --volume-type {volume}' \
+                                                    .format(g=repair_group_name, repair=repair_vm_name, vault=key_vault, kek_url=key_encryption_url, volume=volume_type)
+                    logger.info('Installing extension on repair VM with KEK\n')
+                    _call_az_command(install_ade_extension_command)
+                elif encryption_type == "single_without_kek":
+                    disk_id = source_vm.storage_profile.os_disk.managed_disk.id
+                    disk_settings_command = 'az disk show --ids {ids} -o json' \
+                                            .format(ids=disk_id)
+                    settings = _call_az_command(disk_settings_command)
+                    settings1 = json.loads(settings)
+                    settings2 = (settings1["encryptionSettingsCollection"]["encryptionSettings"])
+                    key_vault = (settings2[0]['diskEncryptionKey']['sourceVault']['id'])
+                    install_ade_extension_command = 'az vm encryption enable --disk-encryption-keyvault {vault} --name {repair} --resource-group {g} --volume-type {volume}' \
+                                                    .format(g=repair_group_name, repair=repair_vm_name, vault=key_vault, volume=volume_type)
+                    logger.info('Installing extension on repair VM without KEK \n')
+                    _call_az_command(install_ade_extension_command)
+            except AzCommandError as azCommandError:
```

1) Contact extension owner and ask why this exception is happening. Have a thread ongoing to remove the exception if nothing is wrong. I still don't completely understand why the exception is happening.
2) Only ignore exception when the exception is "Failed to encrypt data volumes". if not, throw the exception again.

---

1) When we attach a copy of disk as data disk to repair VM we need encryption settings ( passphrase ) to unlock it. Thats the reason we are installing extension on repair VM by choosing volume type as data.

But this extension installation command's primary job is to encrypt the disk. It happens in two phase  i) It pushes the encryption settings ii) It encrypts the disk.

ADE does not support encryption if data disk have xfs file system and hence it fails in the 2nd phase. But since 1st phase is succeeded we will be having passphrase to unlock the disk and hen this error can be ignored. This solution was proposed by ADE SMEs and PG and hence we can ignore this error.

2) I am still working on ignoring "failed to encrypt data disk" message alone.

---

is there any way to install the extension without doing the second phase (failed to encrypt data disk) so that this doesn't happen? Do you know who owns the CLI command for this maybe we can ask that person for more details.

---

Also would the 2nd phase not fail for some situations? for example when the disk does not have xfs file system. What happens then?

---

1) We dont have control over extension installation phases.
2) Yes the second phase does not fail for windows VMs as windows file-system is not XFS.